### ISSUE TITLE
constructor-super - super() must be called even if the class extends null

### ISSUE THREAD

**Issue by nicolo-ribaudo** [eslint/eslint#5449]
**What version of ESLint are you using?**
**What configuration and parser (Espree, Babel-ESLint, etc.) are you using?**

I'm using the online demo with the default configuration (+ `es6`).

**What did you do? Please include the actual source code causing the issue.**
```js
class A extends null {
	constructor() {
		super();
	}
}

new A();
```
```js
class B extends null {
	constructor() {}
}

new B();
```

**What did you expect to happen?**

The first code should be fine, in the second ESLint should warn `2:2 - Expected to call 'super()'.`

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint warns `3:3 - Unexpected 'super()'.` for the first code, while the second is considered fine.

**Specification:**

- In a class definition, if *ClassHeritage* is present (so even if it is `extends null`), the class' *[[ConstructorKind]]* internal slot is set to `"derived"`. <sup>[14.5.14](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-runtime-semantics-classdefinitionevaluation) - step 15</sup>
- When instantiating a new class<sup>[9.2.2 - [[Construct]]](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-ecmascript-function-objects-construct-argumentslist-newtarget)</sup>, *OrdinaryCallBindThis*<sup>[9.2.1.2](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-ordinarycallbindthis)</sup> (which sets *envRec.[[thisBindingStatus]]* to `"initialized"` by calling the *BindThisValue*<sup>[8.1.1.3.1](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-bindthisvalue)</sup>) is only called if `[[ConstructorKind]]` is `"base"`<sup>[9.2.2](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-ecmascript-function-objects-construct-argumentslist-newtarget) - Steps 4 and 8</sup>, so *envRec.[[thisBindingStatus]]* remains `"uninitialized"` even if the class `extends null`.
- The *[[Construct]]* function<sup>[9.2.2](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-ecmascript-function-objects-construct-argumentslist-newtarget) - step 15</sup> returns the result of calling the *envRec.GetThisBinding()* function, which throws if *envRec.[[thisBindingStatus]]* is `"uninitialized"`<sup>[8.1.1.3.4](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-function-environment-records-getthisbinding) - Step 3</sup>.

**Comment by mysticatea** [comment]
Thank you for the issue.

I need to investigate the spec.
In both patterns, Firefox and Chrome throw an error.

**Comment by BYK** [comment]
Tested this on Node 4.3  - The class definition works without any issues but when I try to instantiate the class it throws an error:

>ReferenceError: this is not defined

I think this makes sense so my intention is to close this ticket. That said we should probably consider a new rule, `no-extend-null`, to prevent the problematic case in the first place. @eslint/eslint-team, thoughts?

**Comment by ilyavolodin** [comment]
@BYK We ran into issues before were spec said one thing, but all of the implementations did something else (like `const` inside `for` loop, none of the browsers supported it), and we chose to go with what spec said. So if our rule is doing something that goes against spec, we should fix our rule.

**Comment by BYK** [comment]
@ilyavolodin - if you say so, I guess we can check if the immediate parent is `null` and not warn if that's the case. I still think we should also warn if the parent is `null` and `super()` is used.

**Comment by michaelficarra** [comment]
@BYK eslint cannot always know when the superclass is `null`. Unless you're only considering the case when a `null` literal is used as the superclass.

**Comment by BYK** [comment]
@michaelficarra - yeah, I know :( I was referring to the very limited "immediate parent" case with a `null` literal.

**Comment by mysticatea** [comment]
After I have investigated this, I think this rule should warn `super()` of a class which inherits from `null`.

In `extends null` case, [GetSuperConstructor] throw a TypeError. This is meaning `super()` always throws a TypeError by spec.
On the other hand, there is a case that the constructor of a class which inherits from `null` is valid.

```js
class A extends null {
    constructor() {
        return {hello: 1};
    }
}
```

In this case, we cannot write `super()`.
(though I believe this is an extreme corner case....)

[GetSuperConstructor]: http://www.ecma-international.org/ecma-262/6.0/index.html#sec-getsuperconstructor

----

By the way, I noticed there are other cases similar to `null` : `extends 0`, `extends "foo"`, etc.
So I'm thinking I add them to the special case as same as `null`.

Thought?

**Comment by nzakas** [comment]
> After I have investigated this, I think this rule should warn super() of a class which inherits from null.

Do you mean *should not*?

**Comment by michaelficarra** [comment]
@nzakas No, should.

@mysticatea See [my comment above](https://github.com/eslint/eslint/issues/5449#issuecomment-191792610). I don't think we should report it if it only works for literals.

**Comment by mysticatea** [comment]
@nzakas I meant "should warn". (it means keeping the current behavior). Because `super()` of a class which inherits from `null` always throws a TypeError.
Of course, in lacking `super()` case, the constructor throws a ReferenceError in most cases. But there are cases that the constructor doesn't throw it.

@michaelficarra Yes. Actually, I'm hovering between two:

- I remove the special case for null and undefined.
- I add the special case for more literals.

Honestly, both cases (`super()` exists or not) cause errors, so I don't have strong opinion here.

**Comment by nzakas** [comment]
I think having special cases for literals is fine. We do want to try to warn people about as much as possible before they run code, so if there's something we can easily detect, I think we should do it.

**Comment by mysticatea** [comment]
**Memo**: `extends null` may have a spec bug: https://github.com/tc39/ecma262/issues/543
